pkg/frontend/querymiddleware/experimental_functions.go

```diff
@@ -87,18 +108,45 @@ func (m *experimentalFunctionsMiddleware) Do(ctx context.Context, req MetricsQue
 	return m.next.Do(ctx, req)
 }
 
+func raiseError(tenantID string, name string, t int) error {
+	switch t {
+	case isFunction:
+		err := fmt.Errorf("function %q is not enabled for tenant %s", name, tenantID)
+		return apierror.New(apierror.TypeBadData, DecorateWithParamName(err, "query").Error())
+	case isAggregator:
+		err := fmt.Errorf("aggregate %q is not enabled for tenant %s", name, tenantID)
+		return apierror.New(apierror.TypeBadData, DecorateWithParamName(err, "query").Error())
+	case isExtendedRangeSelector:
+		err := fmt.Errorf("extended range selector modifier %q is not enabled for tenant %s", name, tenantID)
+		return apierror.New(apierror.TypeBadData, DecorateWithParamName(err, "query").Error())
+	}
+	return nil
+}
+
 // containedExperimentalFunctions returns any PromQL experimental functions used in the query.
-func containedExperimentalFunctions(expr parser.Expr) map[string]struct{} {
-	expFuncNames := map[string]struct{}{}
+func containedExperimentalFunctions(expr parser.Expr) map[string]int {
```

Comment:
Rather than returning a `map[string]int` like this, what if we added a `experimentalFeature` struct with name and type fields?